### PR TITLE
Add code action to rename import

### PR DESCRIPTION
**Pull request by tgodzik**
Created a quick WiP to gather some feedback. I found it often that I had to manually add braces and move the cursor to a correct place although it is pretty easy to add this code action.

rename-import

### CODE REVIEW

**Review by dos65 — APPROVED** [review]
That's cool! :rocket:

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/RenameImport.scala [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          }
+        case Some(value) => findLastEnclosingAtPos(value, pos)
+      }
+    }
+
+    def renameImportCodeAction(name: Name): Option[l.CodeAction] =
+      name.parent.flatMap {
+        case nm @ Importee.Name(_) =>
+          val range = nm.pos.toLSP
+          val text = s"{ $name => }"
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/codeactions/RenameImport.scala:46]: So I'm not fully sure how this works with cursor position. For example I tried it in both nvim-metals and coc-metals just to see how two different clients would handle the codeAction edit and the cursor seems to stay in place for both of them.


### nvim-metals




### coc-metals

**Review by ckipp01 — COMMENTED** [review]
Really neat idea! I think we need to figure out the cursor positioning though because on both of the clients I tried it on, the cursor ended up staying in place.

**Comment by tgodzik** [comment]
> Really neat idea! I think we need to figure out the cursor positioning though because on both of the clients I tried it on, the cursor ended up staying in place.

Yeah, I managed to get it working by exposing the command in the extension, but I wanted to get some feedback before trying to solve this.

I wonder if there is any specific way in LSP to make sure that the cursor ends up at a particular place 🤔

**Comment by ckipp01** [comment]
> > Really neat idea! I think we need to figure out the cursor positioning though because on both of the clients I tried it on, the cursor ended up staying in place.
> 
> Yeah, I managed to get it working by exposing the command in the extension, but I wanted to get some feedback before trying to solve this.
> 
> I wonder if there is any specific way in LSP to make sure that the cursor ends up at a particular place 🤔

I don't know if it's possible to do https://github.com/microsoft/language-server-protocol/issues/724

**Comment by tgodzik** [comment]
> > > Really neat idea! I think we need to figure out the cursor positioning though because on both of the clients I tried it on, the cursor ended up staying in place.
> > 
> > 
> > Yeah, I managed to get it working by exposing the command in the extension, but I wanted to get some feedback before trying to solve this.
> > I wonder if there is any specific way in LSP to make sure that the cursor ends up at a particular place thinking
> 
> I don't know if it's possible to do [microsoft/language-server-protocol#724](https://github.com/microsoft/language-server-protocol/issues/724)

We might need to use the command then. Is `GotoLocation` command exposed in the other clients? From what I checked, edit is applied first and then the command executed. So in theory it should work unless nvim or coc do not support running client commands? We could do a workaround:
- code action has an attached command `metals.goto-location`
- serwer gets that command and invokes ClientCommands.GoToLocation
- client executes that command, which should already be supported

What do you think?

**Comment by ckipp01** [comment]
> > > > Really neat idea! I think we need to figure out the cursor positioning though because on both of the clients I tried it on, the cursor ended up staying in place.
> > > 
> > > 
> > > Yeah, I managed to get it working by exposing the command in the extension, but I wanted to get some feedback before trying to solve this.
> > > I wonder if there is any specific way in LSP to make sure that the cursor ends up at a particular place thinking
> > 
> > 
> > I don't know if it's possible to do [microsoft/language-server-protocol#724](https://github.com/microsoft/language-server-protocol/issues/724)
> 
> We might need to use the command then. Is `GotoLocation` command exposed in the other clients? From what I checked, edit is applied first and then the command executed. So in theory it should work unless nvim or coc do not support running client commands? We could do a workaround:
> 
>     * code action has an attached command `metals.goto-location`
> 
>     * serwer gets that command and invokes ClientCommands.GoToLocation
> 
>     * client executes that command, which should already be supported
> 
> 
> What do you think?

Yea in theory I think that should work.

**Comment by tgodzik** [comment]
Some further experiments:

metals-sample-1615827812535

This time I am first adding rename with some random new name, then running the command to save and rename that new symbol. Looks to be working fine aside from the wating time in case of saving the file. We should be able to work around that, but not yet sure how.

**Comment by tgodzik** [comment]
I might get back to this at some point, but closing for now. Not entirely satisfied with what can be done.

**Pull request closed without merging**